### PR TITLE
initial transfer of LG model from gpjet

### PR DESCRIPTION
**Pull request by pgkirsch**

### CODE REVIEW

**Comment by pgkirsch** [comment]
Ready for review @whoburg 

Also, I added a constraint and the result got much closer to the actual 737! Unfortunately, I have no idea why it changed the things it did....

File: aircraft/landing_gear.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    
+    Sources:
+    AERO 481 notes
+    www.fzt.haw-hamburg.de/pers/Scholz/dglr/hh/text_2002_04_11_Fahrwerk.pdf
+    http://faculty.dwc.edu/sadraey/Chapter%209.%20Landing%20Gear%20Design.pdf
+    http://www.dept.aoe.vt.edu/~mason/Mason_f/M96SC03.pdf
+    http://asm.matweb.com/search/SpecificMaterial.asp?bassnum=M434AE
+    http://asm.matweb.com/search/SpecificMaterial.asp?bassnum=M434AE
+    https://en.wikipedia.org/wiki/Buckling
+    http://arc.aiaa.org/doi/pdf/10.2514/3.7771
```
Comment (whoburg) [inline aircraft/landing_gear.py]: cite AIAA-2016-2003 (your scitech paper)

File: aircraft/landing_gear.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        zwing     = Variable('z_{wing}', 0.5, 'm',
+                             'Height of wing relative to base of fuselage')
+
+        with SignomialsEnabled():
+
+            objective = W_lg
+
+            constraints = [
+                           # Track and Base geometry definitions
+                           l_n + zwing + y_m*tangam >= l_m, # [SP]
```
Comment (whoburg) [inline aircraft/landing_gear.py]: yet another example of a Signomial Constraint where as a modeler I know a good initial guess:

a reasonable initial monomial approximation of `l_n + zwing + y_m*tangam` is `l_n`.

File: aircraft/landing_gear.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+                           # Brake sizing for stopping aircraft
+
+                           # Hard landing
+                           # http://www.boeing.com/commercial/aeromagazine/...
+                           # articles/qtr_3_07/AERO_Q307_article3.pdf
+                           # sink rate of 10 feet per second at the maximum
+                           # design landing weight
+                           # Landing condition from Torenbeek p360
+                           Eland == W_0/(2*g)*w**2, # Torenbeek (10-26)
+                           # S_t == 0.5*lam*Lwm/(p*(dtm*bt)**0.5), # (10-30)
```
Comment (whoburg) [inline aircraft/landing_gear.py]: what's up with the commented constraint?

File: aircraft/landing_gear.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                           W_nw >= nwps*WAWn,
+
+                           # Buckling constraint on main gear
+                           L_m <= np.pi**2*E*I_m/(K*l_m)**2,
+                           I_m == np.pi*r_m**3*t_m,
+
+                           # Buckling constraint on nose gear
+                           # source: https://en.wikipedia.org/wiki/Buckling
+                           L_n <= np.pi**2*E*I_n/(K*l_n)**2,
+                           I_n == np.pi*r_n**3*t_n,
```
Comment (whoburg) [inline aircraft/landing_gear.py]: Can `I_n` be converted from `==` to `<=`?

File: aircraft/landing_gear.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+                           2*r_m/t_m <= 40,
+                           2*r_m/t_n <= 40,
+
+                           # Retraction constraint on strut diameter
+                           2*wtm + 2*r_m <= hhold,
+                           2*wtn + 2*r_n <= 0.8*units.m, #TODO improve this
+
+                           # Weight accounting
+                           Waddm == 1.5*W_mw, # Currey p264
+                           Waddn == 1.5*W_nw,
```
Comment (whoburg) [inline aircraft/landing_gear.py]: can't the above two constraints be made into inequalities?

**Comment by whoburg** [comment]
Looks good. a number of equality constraints can be converted to inequalities; otherwise I think this is ready to go.

Also interesting is that for me `LandingGear().test()` takes about 2 seconds with cvxopt, and under one second with mosek.

**Comment by whoburg** [comment]
@bqpd I think we're starting to get enough models that it would make sense to have models-gpkit unit tests that just run all the `test()` methods. Do we already have that in place? How do I run it? Should we ask Marshall to set up Jenkins to run it too?

**Comment by bqpd** [comment]
I'd say import a testing.py file which imports ask the desired models to
top level, then for through dir running everything with a test method.

**Comment by bqpd** [comment]
(smartphone typos, sorry)

**Comment by pgkirsch** [comment]
The only reason I'm hesitant to convert equality constraints to inequalities is because I have a suspicion that it makes them more susceptible to being dual infeasible in the future (if another model is added that puts a stronger pressure in the opposite direction). Do you buy that? The four changes you propose don't affect the result in the present model, but when I debug dual infeasibility, the first thing I do is make all inequality constraints equalities wherever possible, and it can sometimes help.

**Comment by whoburg** [comment]
@pgkirsch, regarding your points about equality constraints, this illustrates to me that we still need a more formal way to talk about the assumptions/pressures of a model. Until we have that I don't think we can issue a blanket policy on when to use equality constraints.

@bqpd, two questions regarding your suggestion to have a `testing.py` file import "the desired models" and then `dir` through everything with a test method:
1. dumb question, but `dir` returns a list of strings. Is it obvious how to get the actual python objects from the strings?
2. Would you do things like `from beam import *`, trying to require less updates to `testing.py`, or would you use explicit imports?

**Comment by bqpd** [comment]
```python

TESTS = []

import beam
TESTS.append(beam)

for attr in dir(beam):
    if hasattr(getattr(beam, attr), "test"):
        getattr(beam, attr).test()  # wrap this in a nice error message

```

**Comment by whoburg** [comment]
Ah - got it. I thought you were suggesting using `dir` to find the modules.

I'll take a shot at adding the testing script today.

**Comment by whoburg** [comment]
`test()` currently fails for me:

```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-3-b8860f1c5b14> in <module>()
----> 1 lg.test()

/Users/whoburg/MIT/dev/models-gpkit/aircraft/landing_gear.py in test(self)
    249     def test(self):
    250         sol = self.localsolve()
--> 251         npt.assert_almost_equal(sol('x_n') + sol('B'), sol('x_m'), decimal=5)
    252         npt.assert_almost_equal(sol('\Delta x_n') + sol('x_n'), sol('x_{CG}'),
    253                                 decimal=1)

/Users/whoburg/anaconda/envs/gpkit/lib/python2.7/site-packages/numpy/testing/utils.pyc in assert_almost_equal(actual, desired, decimal, err_msg, verbose)
    488         pass
    489     if round(abs(desired - actual), decimal) != 0 :
--> 490         raise AssertionError(_build_err_msg())
    491
    492

AssertionError:
Arrays are not almost equal to 5 decimals
 ACTUAL: 21.284771127378246
 DESIRED: 21.284763621703661
```

**Comment by whoburg** [comment]
also, I added the testing script discussed above in #29 -- we can move that conversation there

**Pull request merged**